lib/route/nh_encap_seg6.c

```diff
@@ -0,0 +1,166 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+
+#include "nl-default.h"
+
+#include "nl-route.h"
+#include "nexthop-encap.h"
+#include "seg6.h"
+
+#include <linux/lwtunnel.h>
+#include <linux/seg6_iptunnel.h>
+
+#include <netlink/attr.h>
+#include <netlink/route/nexthop.h>
+
+#include "nl-priv-dynamic-core/nl-core.h"
+
+static int seg6_encap_build_msg(struct nl_msg *msg, void *priv)
+{
+	struct seg6_iptunnel_encap *slwt;
+
+	slwt = priv;
+	return nla_put(msg, SEG6_IPTUNNEL_SRH, SEG6_IPTUN_ENCAP_SIZE(slwt), slwt);
+}
+
+static struct nla_policy seg6_encap_policy[SEG6_IPTUNNEL_MAX + 1] = {
+	[SEG6_IPTUNNEL_SRH] = { .minlen = sizeof(struct seg6_iptunnel_encap) +
+	                                  sizeof(struct ipv6_sr_hdr) +
+	                                  sizeof(struct in6_addr) },
+};
+
+/**
+ * Copied from Linux 6.4: seg6_build_state:net/ipv6/seg6_iptunnel.c
+ * Author: David Lebrun <david.lebrun@uclouvain.be>
+ */
+static int seg6_encap_parse_msg(struct nlattr *nla, struct rtnl_nexthop *nh)
+{
+	struct nlattr *tb[SEG6_IPTUNNEL_MAX + 1];
+	struct seg6_iptunnel_encap *tuninfo;
+	int err, tuninfo_len;
+	struct rtnl_nh_encap *rtnh_encap;
+	struct seg6_iptunnel_encap *slwt;
+
+	err = nla_parse_nested(tb, SEG6_IPTUNNEL_MAX, nla, seg6_encap_policy);
+	if (err < 0)
+		return err;
+
+	if (!tb[SEG6_IPTUNNEL_SRH])
+		return -NLE_INVAL;
+
+	tuninfo = nla_data(tb[SEG6_IPTUNNEL_SRH]);
+	tuninfo_len = nla_len(tb[SEG6_IPTUNNEL_SRH]);
+
+	switch (tuninfo->mode) {
+	case SEG6_IPTUN_MODE_INLINE:
+		break;
+	case SEG6_IPTUN_MODE_ENCAP:
+		break;
+	case SEG6_IPTUN_MODE_L2ENCAP:
+		break;
+	case SEG6_IPTUN_MODE_ENCAP_RED:
+		break;
+	case SEG6_IPTUN_MODE_L2ENCAP_RED:
+		break;
+	default:
+		return -NLE_INVAL;
+	}
+
+	/* verify that SRH is consistent */
+	if (!seg6_validate_srh(tuninfo->srh, tuninfo_len - sizeof(*tuninfo), false))
```

Comment:
why is this a good idea?

Kernel clearly needs to validate, the netlink messages that it gets from user-space. But here we get netlink messages from kernel. We trust kernel to a certain extend to do the right thing, and if it doesn't, shouldn't we try to make the best of it, and interpret it the best we can?

The check like `if (IPV6_EXTHDR_LEN(srh->hdrlen) != len)` is of course important, to ensure that we have the expected buffer size. But why would libnl care about `if (srh->type != IPV6_SRCRT_TYPE_4)` ?

I don't understand `seg6_validate_srh()`? Why is this so complicated? I wouldn't mind, but `rtnl_route_nh_get_encap_seg6_srh()` exposes this complexity to the user (by giving access to that struct, which seems complex to interpret). Is the user expected to re-implment how to interpret a `struct ipv6_sr_hdr`?